klippy/extras/pinda2.py

```diff
@@ -0,0 +1,103 @@
+# P.I.N.D.A. 2 support
+#
+# Copyright (C) 2019  Rafal Jastrzebski <r2@jastrzebscy.org>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+# configuration example:
+#[pinda2]
+#pin: PB4
+#x_offset: 24
+#y_offset: 5
+#z_offset: 0.25
+#speed: 10.0
+#sensor_pin: PF3
+#sensor_type: EPCOS 100K B57560G104F
+#min_temp:0
+#max_temp:100
+#z_offset_calibration:[(0, 0.00),(40, 0.00),(50, -0.25), (100, -0.50)]
+
+#additionl command:
+#PROBE_TEMP
+# gives temperature of P.I.N.D.A. sensor
+import pins, homing, manual_probe, adc_temperature, probe
+
+class Pinda2Probe (probe.PrinterProbe):
+    def __init__(self, config, mcu_probe):
+        probe.PrinterProbe.__init__(self, config, mcu_probe)
+        self.gcode.register_command('PROBE_TEMP', self.cmd_PROBE_TEMP,
+                                    desc=self.cmd_PROBE_TEMP_help)
+
+    cmd_PROBE_TEMP_help = "Return the temperature of P.I.N.D.A 2 sensor"
+    def cmd_PROBE_TEMP(self, params):
+        temp = self.mcu_probe.last_temp
+        self.gcode.respond_info( "pinda2_temp: %s" % temp)
+
+class Pinda2EndstopWrapper(probe.ProbeEndstopWrapper):
+    def __init__(self, config):
+        probe.ProbeEndstopWrapper.__init__(self, config)
+        self.init_sensor(config)
+
+    def init_sensor(self, config):
+        self.speed = config.getfloat('speed', 5.0)
+        self.min_temp = config.getfloat('min_temp', minval=-273)
+        self.max_temp = config.getfloat('max_temp', above=self.min_temp)
+        self.sensor = self.printer.lookup_object('heater').setup_sensor(config)
+        self.sensor.setup_minmax(self.min_temp, self.max_temp)
+        self.sensor.setup_callback(self.temperature_callback)
+        self.printer.lookup_object('heater').register_sensor(config, self)
+
+        s = config.get('z_offset_calibration')
+
+        try:
+            self.z_offset_calibration_samples= eval(s)
+        except SyntaxError:
```

Comment:
It might be a better idea to parse temperature,offset pairs the same way tuples are parsed elsewhere in Klipper rather than using eval() to parse a python style list of tuples.  For example, the configuration would look like the following:

```
# Temperature,Offset Pairs
temperature_offsets:
 30, 0.1
 40, 0.2
 50, 0.3
```

Reply:
I prefer eval(). It opens some tricks like calculation of the correction using the function.
but it is more general discussion.
It could by useful for custom thermistors for instance.

Reply:
I'd prefer not to use eval - it has obscure security and interface issues that complicate maintenance.  For the config file, we should document what each field is (in config/example-extras.cfg) and then parse those fields out.

Reply:
Have you changed your mind about the eval function?
https://github.com/KevinOConnor/klipper/commit/8b005808848dd5395792dba8b9f470ac441c5e72

Reply:
ok, only literals.